spec_util/meld.go

```diff
@@ -478,18 +480,42 @@ func structShouldBeMap(struc *pb.Struct) bool {
 	// A struct should be a map if its number of optional fields exceeds
 	// maxOptionalFieldsPerStruct.
 	numOptionalFields := 0
-	for _, field := range struc.Fields {
+	allFieldsStartWithNumbers := true
+	for fieldName, field := range struc.Fields {
 		if field.GetOptional() != nil {
 			numOptionalFields++
 			if numOptionalFields > maxOptionalFieldsPerStruct {
 				return true
 			}
 		}
+		if !startsWithNumber(fieldName) {
+			allFieldsStartWithNumbers = false
+		}
+	}
+
+	// A struct should be a map if all its fields start with numbers and there
+	// are a sufficient number of fields.  Because many programming languages
+	// disallow struct names starting with numbers, the number of fields is
+	// lower.
```

Comment:
An alternative might be to convert to a map if there are enough field names starting with a number, even in the presence of field names starting with other characters.

Another heuristic might be to look at how many fields have certain kinds of punctuation in their names.

What's here seems fine, though.